Add/translation of hardware and web article (#1)
parabolabam/web-standards.ru: src/articles/hardware-and-web/index.md

---
title: 'Веб и железки'
date: 2021-10-11
author: 'Niels Leenheer'
source:
    title: 'Hardware and the web: the balance between usefulness, security and privacy'
    url: 'https://nielsleenheer.com/articles/2021/hardware-and-the-web-the-balance-between-usefulness-security-and-privacy/'
translators:
    - vlad-sorokin
editors:
    - vadim-makeev
tags:
    - hardware
    - web
preview: 'Веб и железки: баланс между удобством, безопасностью и приватностью'
---
Около года назад, Apple выпустила список Веб API, которые не они не собирались внедрять в Safari из соображений приватности и проблем с безопасностью. Они беспокоились, что эти API позволят создавать цифровой след пользователя, что позволит их отслеживать. И это, конечно же, табу, говоря в терминах приватности. Звучит разумно, верно?

#### Но что за API были в том списке?
 Ну, среди прочих, API для управления железками, которые поставлялись в браузерах Chrome и Edge за последние пару лет:
 - WebBluetooth;
 - WebHID;
 - WebMidi
 - WebNFC;
 - WebSerial;
 - WebUSB.

 Звучит ну прям опасно, да?

>  И я понимаю некоторые сомнения по поводу включения этих API в веб. Они мощные, но потому и кажутся потенциально опасными. Никто не хотет, чтобы случайный веб-сайт захватил контроль над девайсами в вашем доме, доставлял вредоносный код, шпонил за вами или чего жуже. Я понимаю сомнения.

Author: ```suggestion
 И я понимаю некоторые сомнения по поводу включения этих API в веб. Они мощные, но потому и кажутся потенциально опасными. Никто не хотет, чтобы случайный веб-сайт захватил контроль над девайсами в вашем доме, доставлял вредоносный код, шпионил за вами или чего хуже. Я понимаю сомнения.
```

 Но нам следует взглянуть на фактические опасности, рассмотреть их должным образом и не верить всяким "нутром чую" или строить наши мнения на основе того самого эпизода "Черного Зеркала", что мы однажды видели...

>  Реальность такова, что для каждоuj фичи, что предлагает веб платформа существует **баланс между удобством, безопасностью и приватностью**. И эти API определенно удобны. И, что более важно, они также относительно безопасны. Я думаю, что добавление этих API в веб-платформу даже увеличит безопасность пользователей.

Author: ```suggestion
 Реальность такова, что для каждой фичи, что предлагает веб платформа существует **баланс между удобством, безопасностью и приватностью**. И эти API определенно удобны. И, что более важно, они также относительно безопасны. Я думаю, что добавление этих API в веб-платформу даже увеличит безопасность пользователей.
```

>  Как и любой API, подразумеващий безопасность и приватность, веб-сайт не может использовать этот API без согласия полдьзователя. Веб-сайт должен спросить разрешения пользователя, что использовать его веб-камеру, и аналогично, веб-сайт должн спросить разрешения на использование Bluetooth девайса или USB девайса. И разрешение не для API - оно для того, что использовать отдельно взятый девайс с помощью этого API. Веб-сайт не знает какой именно девайс будет использован и не может узнать список таких девайсов. Веб-сайт может только спрашивать разрешения получить доступ к определенному типу девайса, и либо получит либо нет.

Author: ```suggestion
 Как и любой API, подразумеващий безопасность и приватность, веб-сайт не может использовать этот API без согласия пользователя. Веб-сайт должен спросить разрешения пользователя, чтобы использовать веб-камеру, и аналогично, веб-сайт должен спросить разрешения на использование Bluetooth девайса или USB девайса. И разрешение не для API - оно для того, чтобы использовать отдельно взятый девайс. Веб-сайт не знает, какой именно девайс будет использован, и не может узнать список таких девайсов. Веб-сайт может только спрашивать разрешения получить доступ к определенному типу девайса.  Либо получит, либо нет.
```

>  Пользователь всегда должен быть единственным, кто держит свои девайсы под контролем и только он должен решать давать ли доступ к определенному типу девайсов или нет.

Author: ```suggestion
 Пользователь всегда должен быть единственным, кто контролирует свои девайсы. Только он должен решать давать доступ к определенному типу девайсов или нет.
```

## Проблемы с безопасностью

> Главная проблема связана не с технологиями, но с социальной инженерией - вынудить пользователь делать что-то, что каким-то образом вредит его.

Author: ```suggestion
Главная проблема связана не с технологиями, а с социальной инженерией. То есть вынудить пользователя делать что-то, что вредит ему.
```

> Например, вставка вредоносного кода в веб-сайт, который предупреждает пользователей о псевдо-проблеме и вынуждает их подключиться к девайсу, чтобы загрузить драйверы, которые скомпрометируют устройство или, например, будут использовать сторонние каналы для извлечения ценной информации, хранящейся на этом девайсе. Для успешного выполнения такого рода действий всегда необходимо *обманывать* пользователя, потому что, как уже было упомянуто, пользователю всегда приходится давать разрешение на доступ к девайсу перед тем как вредоносный код сможет сам получить доступ к девайсу.

Author: ```suggestion
Например, вставка в веб-сайт вредоносного кода. Он предупреждает пользователей о псевдо-проблеме и вынуждает их подключиться к девайсу, чтобы загрузить драйверы, которые скомпрометируют устройство. Или драйверы будут использовать сторонние каналы для извлечения ценной информации на этом девайсе. Для успешного выполнения такого рода действий всегда необходимо *обманывать* пользователя, потому что всегда приходится давать разрешение на доступ к девайсу перед тем как вредоносный код сможет сам получить его.
```

> Я бы с большим удовольствием сказал бы, что это невозможно, но, к сожалению, это существующая проблема. Социальная инженерия - всегда проблема и не ограничена лишь использованием API.

Author: ```suggestion
Я с большим удовольствием сказал бы, что это невозможно, но, к сожалению, это существующая проблема. Социальная инженерия - всегда проблема и не ограничена лишь использованием API.
```

> Но даже если это теоритически возможно, кажется маловероятным, что эсксплуатироваться это будет на крпных масштабах.

Author: ```suggestion
Даже если это теоритически возможно, то кажется маловероятным, что это будет эсксплуатироваться на крупных масштабах.
```

> Во-первы, вам необходимо найти уязвимый девайс, который принимает драйверы или код, произведенный не для этого конкретного девайса. Большинство девайсов не будут испытывать с этим проблем, но большинство - не все до единого. Во-вторых, у пользователей, на которых направлена атака, должен быть этот девайс. У большинства, скорее всего, просто его не окажется. И в-третьих, еще нужно обмануть пользователей, чтобы они позволили получить доступ к девайсу. Скорее всего, большинство пользователей будут подозрительны и откажутся предоставлять доступ. И наконец, из-за того, что вопросы о предоставлении доступа к девайсу - главные подозреваемые на наличие потенциальной угрозы и незамедлительно пробьют тревогу для пользователей, я не думаю, что такого рода аттака будет рациональна на сколь угодно расширенный период (то есть долгое выполнение вредоносного когда вряд ли будет иметь смысл).

Author: ```suggestion
Во-первых, необходимо найти уязвимый девайс, который принимает драйверы или код, произведенный не для этого конкретного девайса. Большинство из них не будут https://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax испытывать с этим проблем, но не все. Во-вторых, у пользователей, на которых направлена атака, должен быть этот девайс. У большинства, скорее всего, его не окажется. И в-третьих, еще нужно обмануть пользователей, чтобы они дали доступ к девайсу. Скорее всего, большинство будут осторожными и откажутся предоставлять доступ. И наконец, из-за того, что запросы о предоставлении доступа к девайсу - главные подозреваемые на наличие потенциальной угрозы, то они незамедлительно насторожат пользователей. Я не думаю, что такого рода аттака будет рациональна на сколь угодно долгий период.
```

> Но когда речь заходит о достаточно большой группе пользователей, число девайсов, которые можно скомпроментировать не будет равно нулю.

Author: ```suggestion
Но когда речь заходит о достаточно большой группе пользователей, число девайсов, которые можно скомпроментировать, не будет равно нулю.
```

> Есть простые контр-меры, используемые создателями браузеров для большего сокращения такого риска.

Author: ```suggestion
Есть простые контр-меры, используемые создателями браузеров для уменьшения такого риска.
```

Например:

> - блокирование доступа к специфичным девайсам, у которых известны уязвимости;

Author: ```suggestion
- блокировка доступа к специфичным девайсам, у которых известны уязвимости;
```

-  запрет доступа к целому классу девайсов с проблемами;

> - Удаленноt прекращение работы API, когда его активно используют для эксплуатации уязвмостей в девайсе.

Author: ```suggestion
- Удаленное прекращение работы API, когда его активно используют для эксплуатации уязвмостей в девайсе.
```

На текущий момент браузеры используют все эти контр-меры, но риск все еще не нулевой.

Насколько мне известно, удаленное выключение API использовалось лишь однажды. И то, не потому что API активно эксплуатировалось, а потому что сканы безопасности сообщили об уязвимости. Все равно было не круто использовать удаленное отключение.

> Возвращаясь в 2018, было возможно [вытащить ключ из девайса под названием Yubico U2F device](https://www.yubico.com/support/security-advisories/ysa-2018-02/), используя WebUSB для обходя проверку происхождения запросов, которую браузеры обычно делют. После того как об этом сообщили, Google незамедлительно отключил WebUSB начисто и выпустил обновление, которое сново позволяло включить WebUSB с добавлением всех всех девайсов Yubico в черный список (или список заблокированных устройств).

Author: ```suggestion
Возвращаясь в 2018 год, было возможно [вытащить ключ из девайса под названием Yubico U2F device](https://www.yubico.com/support/security-advisories/ysa-2018-02/), используя WebUSB для обхода проверки происхождения запросов. Такую проверку обычно делают браузеры. После того как об этом сообщили, Google незамедлительно отключил WebUSB и выпустил обновление, которое позволяло включить WebUSB с добавлением всех девайсов Yubico в черный список (или список заблокированных устройств).
```

> Мне кажется, что причина по кторой мы не видели подобных аттак состоит в том, что намного проще загрузить нативное приложение и выполнить его на устрйостве пользователя, чем найти тот самый из миллиона вариантов, в котором будут все составляющее удачного совершения злодеяния: пользователь, которого легко ввести в заблуждение и уязвимый девайс. Проще говоря, игра не стоит свеч.

Author: ```suggestion
Мне кажется, что причина, по которой мы не видели подобных аттак состоит в том, что намного проще загрузить нативное приложение и выполнить его на устройстве пользователя, чем найти тот самый из миллиона вариантов, в котором будут все составляющее удачного совершения злодеяния: пользователь, которого легко ввести в заблуждение и уязвимый девайс. Проще говоря, игра не стоит свеч.
```

Не только Google думает, что API, о которых идет речь, достаточно безопасны, чтобы выкатывать их к реальным пользователям. Того же мнения придерживаются Microsoft и Samsung. В действительности, эти API поставляются на 70% мобильных браузеров и 78% десктопных браузеров по всему миру. И поствляются уже достаточно давно.

> Ноя я думаю, что настоящая проблема безопасности не браузере с этими функциями или без них.

Author: ```suggestion
Но я думаю, что настоящая проблема безопасности не в браузере с этими функциями или без них.
```

> Чтобы выполнить специфичную задачу, у пользователя есть выбор между использованием нативного приложения или браузера. Выбор стоит между использованием ограниченного API, построенного вогруг пользовательского соглашения в изолированной среде и нативным приложением, которое может может делать что угодно, невзирая на приватность или безопасность. Вы бы загрузили какое-то сомнительное приложение, созданное неизвестным разработчиком из магазина приложений, котрое делает кто-знает-что с вашими данными в фоновом режиме?

Author: ```suggestion
Чтобы выполнить специфичную задачу, у пользователя есть выбор между использованием нативного приложения или браузера. Выбор стоит между использованием ограниченного API, построенного вокруг пользовательского соглашения в изолированной среде, и нативным приложением, которое может может делать что угодно, невзирая на приватность или безопасность. Вы бы загрузили какое-то сомнительное приложение, созданное неизвестным разработчиком из магазина приложений, которое делает кто-знает-что с вашими данными в фоновом режиме?
```

Ограничивая API устройства нативными прилоржениями, вы заставляете людей пользоваться нативными приложениями для их задач. И это не хорошо для безопасности.

**Самая опасная функциональность**, которая есть в браузерах - это не API девайсов. Это - **возможность ссылаться и загружать нативные приложения**.

Эта функциональность активно эксплуатируется и эксплуатировалась с самого момента, когда самый первый браузер попал в руки широкой публики. И время доказало, что почти ничего нет, чем мы смогли бы снизить риски.

## Приватность и цифровой след (фингерпринтинг)

Что насчет цифрового следа? Возможно ли проследить цифровой след пользователя, используя эти API?

> По правде говоря, в этом нет много смысла. Исползьвание фингерпринтинга в качестве аргумента против этих API показывает недостаточность понимания как какой-либо из этих API вообще работает.

Author: ```suggestion
По правде говоря, в этом нет много смысла. Использование фингерпринтинга в качестве аргумента против этих API показывает нехватку понимания как какой-либо из этих API вообще работает.
```

> Цель фингерпринтинга - идентифицировать уникального пользователя. Этого можно добиться, используя уникальные опорные данные для большиз групп пользователей. Каждый набор таких данных не может идентифицировать конкретного пользователя, но может оказаться возможным отслеживать пользователей, если у вас будет достаточно опорных данных и из комбинации.

Author: ```suggestion
Цель фингерпринтинга - идентифицировать уникального пользователя. Этого можно добиться, используя уникальные опорные данные для больших групп пользователей. Каждый набор таких данных не может идентифицировать конкретного пользователя, но его можно отследить, если у вас будет достаточно опорных данных и их комбинации.
```

> И эти опорные данные могут быть ничем примечательным, о чем стоило бы беспокоится, говоря о приватности. Или даже те штуки, которые по иронии предназначаются для улучшения приватности. Например, HTTP заголовок "Do Not Track". Он выключен по-умолчанию, поэтому это идеальная вводная для отслеживания пользователей. Если у вас есть пул из 1 миллиона пользователей и один из сотни включил HTTP заголовок "Do Not Track", то считайте, что вы можете наблюдать за одним из 10 тысяч пользователей вместо одного из миллиона. Теперь попробуйте скомбинировать это с еще хотя бы 30 типами невинно выглядящих опорных данных и вы можете отслеживать конкретных пользователей на разных веб-сайтах - святой грааль отслеживания польлзователя.

Author: ```suggestion
Такие данные могут быть ничем примечательным, о чем стоило бы беспокоится, говоря о приватности. Или даже те штуки, которые по иронии предназначаются для улучшения приватности. Например, HTTP заголовок "Do Not Track". Он выключен по-умолчанию, поэтому это идеальная отправная точка для отслеживания пользователей. Например, если у вас есть пул из 1 миллиона пользователей и один из сотни включил использование HTTP заголовка "Do Not Track", то считайте, что вы можете наблюдать за одним из 10 тысяч пользователей вместо одного из миллиона. Теперь попробуйте скомбинировать это с еще хотя бы 30 типами невинно выглядящих опорных данных и вы можете отслеживать конкретных пользователей на разных веб-сайтах - святой грааль отслеживания пользователя.
```

HTTP заголовок "Do Not Track" был удален из Safari как раз из соображений выше. И поделом ему.

Вопрос вот в чем: добавляют ли эти API, перечисленные в начале, опорных данных для отслеживания конечных пользователей?

> Теоретически, да. Простое существование таких API уже может считаться отправной точкой для отслеживания. Но на практике - нет. Возьмем, например, WebBluetooth. Эта технология была доступна в браузере Chrome со времен 56 версии. Поэтому, единственное о чем наличие этого API сигнализирует, так это то, старше ли браузер Chrome 56ой версии или нет. И эту информацию можно добыть просто заглянув в useragent строчку, которая всегда присутствует в HTTP запросе. Поэтому эти данные выглядят крайне бесполезными для фингерпринтинга.

Author: ```suggestion
Теоретически, да. Простое существование таких API уже может считаться отправной точкой для отслеживания. Но на практике - нет. Возьмем, например, WebBluetooth. Эта технология была доступна в браузере Chrome со времен 56ой версии. Поэтому, единственное о чем наличие этого API сигнализирует, так это то, старше ли браузер Chrome 56ой версии или нет. И эту информацию можно добыть, просто заглянув в `useragent` строчку, которая всегда есть в HTTP запросе. Поэтому эти данные выглядят крайне бесполезными для фингерпринтинга.
```

> Для фингерпринтинга нужно API, которое дает разные результаты на разных машинах пользователей. Например, шрифты, установленные польователем на компьютере или GL расширения, которые поддерживает их графическая карта. Эти данные могут отличаться на девайсах, на которых есть установлены две одинаковые версии браузера. Такие опорные данные добавляют информации или энтропию в фингерпринтинг.

Author: ```suggestion
Для фингерпринтинга нужно API, дающее разные результаты на разных машинах пользователей. Например, шрифты, установленные польователем на компьютере или GL расширения, поддерживаемые графической картой. Эти данные могут отличаться на девайсах, на которых установлены две одинаковые версии браузера. Такие опорные данные добавляют энтропии в фингерпринтинг.
```

> Но что все-таки с API этих девайсов? Предлагают ли они какую-то дополнительную информацию полезню для фингерпринтинга? Ну, если браузер знает какие USB девайсы соединены с вашей системой и какой Bluetooth девайс подключен к системе, я бы сказал "Да, абсолютно". **Но эти API работают не так**. Их проектировали, зная о фингерпринтинге, поэтому они не могут использоваться для создания цифрового следа пользователя напрямую.

Author: ```suggestion
Но что все-таки с API этих девайсов? Предлагают ли они какую-то дополнительную информацию полезную для фингерпринтинга? Если браузер знает какие USB девайсы соединены с вашей системой и какой Bluetooth девайс подключен к системе, я бы сказал: "Да, абсолютно". **Но эти API работают не так**. Их проектировали, зная о фингерпринтинге, поэтому они не могут использоваться для этого напрямую.
```

*Нельзя просто так взять и* получить список девайсов вокруг вас. Нельзя так сделать ни с помощью WebBluetooth, ни с помощью чего-либо еще. Эта информация *недоступна* веб-сайтам. Вы можете спать спокойно, зная, что веб-сайты не могут узнать о ваших девайсах, а значит не могут идентифицировать вас через них.

Что веб-сайты могут, так это сказать браузеру с какими девайсами он хотел бы общаться. Обычно, для API предоставляется набор фильтров, основанных либо на имени девайса, либо на предлагаемых девайсом сервисов.

> Браузер затем показывает попап, спрашивающий разрешение, в котором предоставлен список девайсов, указанных в списке фильтров, предоставленным вами. Этот список доступен только самому пользователю, ни один скрипт, исполняющийся в браузере не сможет получить доступ к этой информации. Пользователь может также предоставить доступ определенному девайсу или щакрыть доступ для всех девайсов.

Author: ```suggestion
Браузер затем показывает попап, спрашивающий разрешение. В попапе есть список девайсов, указанных в списке фильтров. Этот список доступен только самому пользователю, ни один скрипт, исполняющийся в браузере, не сможет получить доступ к этой информации. Пользователь может также дать доступ к определенному девайсу или закрыть доступ ко всем.
```

> Этот попап с разрешением позволяет убедиться в том, что кто-то плохой не будет использовать доступ к девайсу для фингерпринтинга, потому что обычно вредоносный код не пытается дать пользователю знать о своих действиях.

Author: ```suggestion
Этот попап позволяет убедиться в том, что кто-то плохой не будет использовать доступ к девайсу для фингерпринтинга, потому что обычно вредоносный код не пытается дать знать о своих действиях.
```

**API девайсов просто плохи для фингерпринтинга**. Это ненадежно и слишком очевидно, когда используется.

## Так что там с Safari...

> Это вполне приемлемо иметь сомнения о безопасноти всех этих API для работы с железками. Лично я думаю, что риски относительно малы и с ними можно справиться. Но также это абсолютно нормально дискутировать или даже не соглашаться с этим. Но указывать на фингерпринтинг и отслеживание будет просто означать, что у вас не совсем корректная информация.

Author: ```suggestion
Это вполне приемлемо иметь сомнения о безопасноти всех этих API для работы с железками. Лично я думаю, что риски относительно малы и с ними можно мириться. Но также, это абсолютно нормально не соглашаться с этим. Но указание на фингерпринтинг и отслеживание как аргумент против использования этих API будет просто означать, что у вас не совсем корректная информация.
```

> Я не совсем против, что Safari не будет добавлять эти API. Разработчики браузеров нуждаются в оценке баланса между удобством, безопасностью и приватностью, чтобы увидеть стоят ли риски этого. Apple думает, что добавление этих API не стоит того. Я так не думаю, но это нормально.

Author: ```suggestion
Я не совсем против, что Safari не будет добавлять эти API. Разработчики браузеров нуждаются в оценке баланса между удобством, безопасностью и приватностью, чтобы увидеть стоят ли риски этого. Apple думает, что добавление этих API не стоит того. Я с этим не согласен, но это нормально.
```

> Я уверен, что они игнорируют риски безопасности перенаправления пользователей в нативные приложения. Но, конечно, я понимаю почему Apple не видит в этом проблемы.

Author: ```suggestion
Уверен, что они игнорируют риски безопасности перенаправления пользователей в нативные приложения, но понимаю почему Apple не видит в этом проблемы.
```

> Чему я возразил бы, так это последствиям этого выбора. Из-за того, что [Safari единственный разрешенный браузер на IOS](https://nielsleenheer.com/articles/2021/chrome-is-the-new-safari-and-so-are-edge-and-firefox/), пользователи не могут выбрать другой браузер, поддерживающий API девайсов. Это большая проблема. Выбор Safari не просто выбор, который негативно сказывается на самом браузере, но также негативно сказывается на всех браузерах на супер важной платформе. И это негативно сказывается на мне как на пользователе.

Author: ```suggestion
Из-за того, что [Safari - единственный разрешенный браузер на IOS](https://nielsleenheer.com/articles/2021/chrome-is-the-new-safari-and-so-are-edge-and-firefox/), пользователи не могут выбрать другой браузер, поддерживающий API железок. Это большая проблема. Выбор Safari - не просто выбор, который негативно сказывается на самом Safari, но и на всех браузерах в целом. И мне от того становится грустно.
```
